Add `make switch` fore easy installation of dependencies

aqjune opened the pull request:
This patch adds `make switch` that
- Creates a local OPAM switch under the current hol-light directory
- Chooses the latest fully supported OCaml version (4.14 for now; would be great if it is 5.0 in the future!)
- And installs dependencies that are needed by HOL Light.

This is to help beginners set the environment for HOL Light.

I also updated README to explain(recommend) this option. A separate 'DEPENDENCIES' chapter is added to explain the original, detailed steps.

aqjune commented on README:
```diff
@@ -24,23 +24,10 @@ Refer to the reference manual for more details of individual functions:
-thanks to Hendrik Tews, so installation of HOL Light and all its
-prerequisites is as simple as
-
-  sudo apt-get install hol-light
```
I found that the apt package of hol-light is not working anymore; should we remove this (as edited in this patch)?

jrh13 replied:
Yes, I agree that we should probably delete this, unless someone volunteers to take over maintenance of the apt package. Direct HOL Light installation (thanks to this patch, among others) is now so much easier anyway.

jrh13 commented:
I have tried this out on several machines and it mostly worked perfectly, thanks! One small suggestion:
On Mac OS I found that my opam complained about "ambiguity" (between ocaml-base-compiler and ocaml-system) in
`opam switch create . 4.14.0`, so I replaced it by `opam switch create . ocaml-base-compiler.4.14.0`, which seemed to work.

I still had trouble on an ARM-based Mac OS, apparently around the use of `m4` in the opam installation process(?) However, I don't really understand the core problem since `m4` should be installed (via homebrew). Probably not worth worrying about for now.

jrh13 commented:
OK, I figured out the m4 problem, which was just the fact that the homebrew version of m4 wasn't in my PATH. So it all worked fine on the ARM-based machine too. So the only suggestion would be the "ocaml-base-compiler" change if you agree.

aqjune commented:
> On Mac OS I found that my opam complained about "ambiguity" (between ocaml-base-compiler and ocaml-system) in
opam switch create . 4.14.0, so I replaced it by opam switch create . ocaml-base-compiler.4.14.0, which seemed to work.

The two seemed identical to me because `opam list` printed identical lists of installed packages. I updated this patch to use `ocaml-base-compiler.4.14.0`.

jrh13 commented:
Great, thank you I've already found it useful on an old machine where my system OCaml was in a weird state. I'll merge this in now.

monadius commented on Makefile:
```diff
@@ -37,6 +37,15 @@ CAMLP5_VERSION=`camlp5 -v 2>&1 | cut -f3 -d' ' | cut -f1-3 -d'.' | cut -f1 -d'-'
+switch:; \
+  opam switch create . ocaml-base-compiler.4.14.0 ; \
+  eval $(opam env) ; \
+  opam install -y zarith camlp5 ledit
```
The `camlp5` version should be explicitly given: `camlp5.8.02.01`. The current version of `camlp5` (8.03) is not compatible with `pa_j_4.xx_8.02.ml`.

aqjune commented:
A problem is that in some machine 8.02 is not available. Does pinning it to
8.00.(some minor version) work?

On Sat, May 25, 2024 at 9:04 PM Alexey Solovyev ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In Makefile
> <https://github.com/jrh13/hol-light/pull/98#discussion_r1614998429>:
>
> > @@ -37,6 +37,15 @@ CAMLP5_VERSION=`camlp5 -v 2>&1 | cut -f3 -d' ' | cut -f1-3 -d'.' | cut -f1 -d'-'
>
>  default: update_database.ml pa_j.cmo hol.sh;
>
> +# Create a local OPAM switch and install dependencies on it.
> +# This will use the latest OCaml version that fully supports features of
> +# HOL Light.
> +# ledit is installed for line editing of OCaml REPL
> +switch:; \
> +  opam switch create . ocaml-base-compiler.4.14.0 ; \
> +  eval $(opam env) ; \
> +  opam install -y zarith camlp5 ledit
>
> The camlp5 version should be explicitly given: camlp5.8.02.01. The
> current version of camlp5 (8.03) is not compatible with pa_j_4.xx_8.02.ml.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jrh13/hol-light/pull/98#pullrequestreview-2079335804>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRPHGKHMTU2RSHRG7CCF6TZEE7JZAVCNFSM6AAAAABGRBKTAGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANZZGMZTKOBQGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

monadius commented:
I tested Camlp5 8.00.03 and it works with OCaml 4.14.
